pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -186,13 +186,15 @@ impl Timeline {
     async fn compact_shard_ancestors(
         self: &Arc<Self>,
         rewrite_max: usize,
-        _ctx: &RequestContext,
+        ctx: &RequestContext,
     ) -> anyhow::Result<()> {
         let mut drop_layers = Vec::new();
-        let layers_to_rewrite: Vec<Layer> = Vec::new();
+        let mut layers_to_rewrite: Vec<Layer> = Vec::new();
 
-        // We will use the PITR cutoff as a condition for rewriting layers.
-        let pitr_cutoff = self.gc_info.read().unwrap().cutoffs.pitr;
+        // We will use the Lsn cutoff of the last GC as a threshold for rewriting layers: if a
+        // layer is behind this Lsn, it indicates that the layer is being retained beyond the
+        // pitr_interval, for example because a branchpoint references it.
+        let last_gc_cutoff = self.gc_info.read().unwrap().cutoffs.pitr;
```

Comment:
As noted in https://github.com/neondatabase/neon/pull/7531/files#r1602860120 and after, instead let's use `Timeline::latest_gc_cutoff_lsn.read()` guard for the whole operation, which will stall gc.

Reply:
We cannot only use `latest_gc_cutoff_lsn` because that is rarely updated (1/h) compared to pitr_cutoff (6/h). I still think we should hold on to a read guard from `latest_gc_cutoff_lsn` regardless, so that a gc cannot happen while rewriting is ongoing.